# Introduction to bioinformatics

## Ribo-seq data analysis

### Description

This part of the practical contains material for the processing and analysis of the *Ribo-seq* data from the example dataset.
The content of the lecture is described below in more detail with learning objectives. We will cover aspects of each section by means of
[Jupyter notebooks](https://jupyter.org/). These notebooks constitute the material for this part of the practical.

### Dependencies

We will *run* the notebooks on our [JupyterHub server](https://jupyter.internal), which provides an access to our computational environments and resources.
    
### Introduction to Ribo-seq data analysis

#### Sections:
    1.1 Short introduction to Jupyter (Notebook) and the JupyterHub
    1.2 Overview of the rpbp pipeline: command line options and configuration file
    1.3 High-level introduction to Ribo-seq, de novo ORF discovery (elements of annotation, transcript isoforms, CDS, UTRs, etc.), biological relevance of alternative translation events (including translation from non-coding transcripts), and why we need "dedicated software" to analyse Ribo-seq data.
    1.4 Ribo-seq quality control how-to.

#### Questions & Objectives:
    - Jupyter Notebook basics: dashboard, user interface, navigation, running code, etc. How to use Jupyter with virtual environments.
    - How to use the JupyterHub.
    - What is the translatome? What are the uses of Ribo-seq.
    - Why do we need dedicated software to analyse Ribo-seq data? What softwares are available to analyse Ribo-seq data?
    - Run one such software, rpbp, on a selected Ribo-seq dataset.
    - Learn how to identify "good quality" Ribo-seq data.
    - Learn how to visualise the results.

#### After I will be able to:
    - Understand what is a Jupyter Notebook, and how to use it.
    - Understand how to analyse Ribo-seq data for ORF discovery.
    - Use the ORF predictions for follow-up studies.

    
